### PR TITLE
Update to syn-2

### PR DESCRIPTION
**Pull request by maurer**
Also moves a test out of doctest compilefail since syn-2 will build it correctly even without the "full" feature.

### CODE REVIEW

**Comment by cuviper** [comment]
Do we use anything that would need "full" anymore? If not, I think we should just drop that feature. We can bump semver for the new MSRV requirement while we're at it, so a breaking feature change is ok.

File: src/lib.rs [hunk 1]
```diff
@@ -170,16 +170,14 @@ impl NumTraits {
...
-                    if let syn::Lit::Str(lit_str) = mnv.lit {
-                        return NumTraits {
-                            import: syn::Ident::new(&lit_str.value(), lit_str.span()),
-                            explicit: true,
-                        };
-                    } else {
-                        panic!("#[num_traits] attribute value must be a str");
-                    }
+            if attr.path().is_ident("num_traits") {
+                if let syn::Expr::Lit(syn::ExprLit { lit: syn::Lit::Str(ref lit_str), .. }) = attr.meta.require_name_value().unwrap().value {
```
Comment (cuviper) [inline src/lib.rs]: I'd rather this fell down to the panic message instead of `unwrap`.

Comment (maurer) [reply src/lib.rs]: My bad, this one slipped through.

**Comment by maurer** [comment]
I fixed up the `.unwrap()` call and added an additional commit to remove the `full-syntax` feature. There are still some things that are not present in `syn` without `full`, but none of them seem to correspond to complex discriminants.

**Comment by glandium** [comment]
@cuviper is there anything holding this up?

**Comment by cuviper** [comment]
> is there anything holding this up?

Lack of time and attention... but it's published now!

**Pull request merged**